hedera-node/docs/design/services/smart-contract-service/frictionless-airdrops-system-contracts.md

```diff
@@ -0,0 +1,151 @@
+
+## References
+
+[HIP-904](https://hips.hedera.com/hip/hip-904) - HIP that introduces the frictionless airdrops.
+
+## Goals
+
+- Expose `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` as new functions in the Hedera Token Service Smart Contract.
+- Expose `cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFt`, `rejectTokenNFTs`, `setAutomaticAssociations` as new functions in the proxy redirect token facade contract IHRC.
+- Implement the needed HTS system contract classes to support the new functions.
+
+## Non Goals
+
+- The implementation of the HAPI operation, as it is already an existing feature.
+
+## Architecture
+
+The architecture for the frictionless airdrops follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.
+
+## Implementation
+
+### New Solidity Structures
+
+We will introduce the following new structures to support the new functionality:
+
+`PendingAirdrop` - A struct that represents a pending airdrop request.
+
+```solidity
+struct PendingAirdrop {
+    address sender;
+    address receiver;
+
+    address token;
+    bool isNft;
+    int64 serial;
+}
+```
+
+`NftId` - A struct that represents the Nft serials to be rejected.
+
+```solidity
+struct NftId {
+    address nft;
+    int64[] serials;
+}
+```
+
+### New Solidity Functions
+
+New system contract functions must be added to the `IHederaTokenService` interface to support airdropping tokens.
+
+| Function Selector Hash |                                                 Function Signature                                                  |    Response    |                                 |
+|------------------------|---------------------------------------------------------------------------------------------------------------------|----------------|---------------------------------|
+| `0x2f348119`           | `function airdropTokens(TokenTransferList[] memory tokenTransfers) external returns (int64 responseCode)`           | `ResponseCode` | The response code from the call |
+| `0xc1acfe5d`           | `function cancelAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`            | `ResponseCode` | The response code from the call |
+| `0x241b6af7`           | `function claimAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`             | `ResponseCode` | The response code from the call |
+| `0x012ea0b1`           | `function rejectTokens(address[] memory ftAddresses,  NftId[] memory nftIds) external returns (int64 responseCode)` | `ResponseCode` | The response code from the call |
```

Comment:
(Oddly, `rejectTokens` does _not_ have a limit (in the HIP) on how many you can reject at one time.  Not sure why you're limited to 10 for cancel but no limit for reject - you'd think similar limits on resource consumption in a single transaction would apply.)